### PR TITLE
Convert annotation

### PR DESCRIPTION
**Pull request by Bhuvanesh-Verma**
This PR adds `convert_annotation.py` which acts as an entry point for the interaction between the span annotations and the tag sequence. This interaction is performed in the following way:
1. `convert_span_annotations_to_tag_sequence` converts the span annotations into the labeled spans and then uses `token_spans_to_tag_sequence` (from `encoding.py`) to obtain a tag sequence. 
2. `convert_tag_sequence_to_span_annotations` uses `tag_sequence_to_token_spans` (from `encoding.py`) to obtain the token spans and then convert the token spans into the labeled spans.

### CODE REVIEW

**Review by ArneBinder — CHANGES_REQUESTED** [review]
Thanks, looks already very good, especially the doc strings are very good to read.

File: src/pie_utils/sequence_tagging/convert_annotation.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+
+
+TypedSpan = Tuple[int, Tuple[int, int]]
+TypedStringSpan = Tuple[str, Tuple[int, int]]
+
+
+def convert_tag_sequence_to_span_annotations(
+    tag_sequence: List[str],
+    token_offset_mapping: List[Tuple[int, int]],
+    offset: int,
```
[the remark this answers is not in the document]
Comment (Bhuvanesh-Verma) [reply src/pie_utils/sequence_tagging/convert_annotation.py]: fixed in [e2d229e](https://github.com/ArneBinder/pie-utils/pull/29/commits/e2d229ef4a0b0cdf9a15da6c15c122992c439b23)

File: src/pie_utils/sequence_tagging/convert_annotation.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+    """Given a tag sequence corresponding to a coding scheme (IOB2, BIOUL and BOUL), this method
+    converts it into the labeled spans.
+
+    # Parameters
+        tag_sequence : `List[str]`, required.
+           The tag sequence encoded in IOB2, BIOUL or BOUL, e.g. ["B-PER", "O", "L-PER"].
+        token_offset_mapping: List[Tuple[int, int]], required.
+            List of tuples containing start and end indices at character level for each token in the tag sequence.
+        offset: int, required.
+            offset to adjust the spans. It is either start index of the partition containing the tag sequence or 0.
```
[the remark this answers is not in the document]
Comment (Bhuvanesh-Verma) [reply src/pie_utils/sequence_tagging/convert_annotation.py]: fixed in [e2d229e](https://github.com/ArneBinder/pie-utils/pull/29/commits/e2d229ef4a0b0cdf9a15da6c15c122992c439b23)

File: src/pie_utils/sequence_tagging/convert_annotation.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+
+
+def span_annotations_to_labeled_spans(
+    span_annotations: BaseAnnotationList[LabeledSpan],
+    char_to_token_mapper: Callable[[int], Optional[int]],
+    partition: Optional[Span] = None,
+    statistics: Optional[DefaultDict[str, Counter]] = None,
+) -> List[Tuple[str, Tuple[int, int]]]:
+    """Given a list of span annotations, a character position to token mapper (as obtained from
+    batch_encoding.char_to_token), create a list of labeled spans.If a partition is provided, only
```
[the remark this answers is not in the document]
Comment (Bhuvanesh-Verma) [reply src/pie_utils/sequence_tagging/convert_annotation.py]: fixed in [e2d229e](https://github.com/ArneBinder/pie-utils/pull/29/commits/e2d229ef4a0b0cdf9a15da6c15c122992c439b23)

File: src/pie_utils/sequence_tagging/convert_annotation.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+            token_offset_mapping[start][0] + offset,
+            token_offset_mapping[end][1] + offset,
+            label,
+        )
+        for label, (start, end) in spans
+    ]
+
+
+def span_annotations_to_labeled_spans(
+    span_annotations: BaseAnnotationList[LabeledSpan],
```
[the remark this answers is not in the document]
Comment (Bhuvanesh-Verma) [reply src/pie_utils/sequence_tagging/convert_annotation.py]: fixed in [e2d229e](https://github.com/ArneBinder/pie-utils/pull/29/commits/e2d229ef4a0b0cdf9a15da6c15c122992c439b23)

File: src/pie_utils/sequence_tagging/convert_annotation.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+) -> List[str]:
+    """Given a list of span annotations, a character position to token mapper (as obtained from
+    batch_encoding.char_to_token) and a special tokens mask, create a sequence of tags with the
+    length of the special tokens mask.
+
+    # Parameters
+        span_annotations : `BaseAnnotationList[LabeledSpan]`, required.
+            List of span annotations
+        base_sequence_length: int, required.
+            length of special token mask
```
[the remark this answers is not in the document]
Comment (Bhuvanesh-Verma) [reply src/pie_utils/sequence_tagging/convert_annotation.py]: fixed in [e2d229e](https://github.com/ArneBinder/pie-utils/pull/29/commits/e2d229ef4a0b0cdf9a15da6c15c122992c439b23)

File: tests/sequence_tagging/test_convert_annotation.py [hunk 6]
```diff
@@ -0,0 +1,440 @@
...
+
+
+@pytest.fixture
+def char_to_token_mappings():
+    # This is created using tokenizer
+    # TEXTS = [TEXT1, TEXT2]
+    # encodings = tokenizer(TEXTS, return_offsets_mapping=True)
+    # result is slightly different from shown below as command below might create some None values
+    # result = [{char_idx: encodings[i].char_to_token(char_idx) for char_idx in range(len(text))} for i, text in enumerate(TEXTS)]
+    return [
```
[the remark this answers is not in the document]
Comment (Bhuvanesh-Verma) [reply tests/sequence_tagging/test_convert_annotation.py]: fixed in [37d1d0f](https://github.com/ArneBinder/pie-utils/pull/29/commits/37d1d0f52f91ae56c9e510e2e0283288c2452130)

File: tests/sequence_tagging/test_convert_annotation.py [hunk 7]
```diff
@@ -0,0 +1,440 @@
...
+def special_tokens_masks():
+    return [
+        [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1],
+        [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1],
+    ]
+
+
+@pytest.fixture
+def true_tag_sequences():
+    return {
```
[the remark this answers is not in the document]
Comment (Bhuvanesh-Verma) [reply tests/sequence_tagging/test_convert_annotation.py:158]: fixed in [37d1d0f](https://github.com/ArneBinder/pie-utils/pull/29/commits/37d1d0f52f91ae56c9e510e2e0283288c2452130)

File: src/pie_utils/sequence_tagging/convert_annotation.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ArneBinder) [reply src/pie_utils/sequence_tagging/convert_annotation.py]: please adjust as discussed

File: src/pie_utils/sequence_tagging/convert_annotation.py [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+
+
+def convert_tag_sequence_to_span_annotations(
+    tag_sequence: List[str],
+    token_offset_mapping: List[Tuple[int, int]],
+    offset: int,
+    encoding: str = "IOB2",
+    include_ill_formed: bool = True,
+    classes_to_ignore: List[str] = None,
+):
```
[the remark this answers is not in the document]
Comment (ArneBinder) [reply src/pie_utils/sequence_tagging/convert_annotation.py]: please adjust as discussed

**Review by ArneBinder — APPROVED** [review]

**Pull request merged**